### ISSUE TITLE
How to check for holidays within an interval of time

### ISSUE THREAD

**Issue by dsoprea** [vacanza/python-holidays#1825]
I'd like to be able to determine if there are any holidays in the near future, expressed as a `timedelta`. More generally, maybe just between two dates.

Pseudocode:
```
NOW_DATETIME = datetime.datetime.utcnow()
SIXTY_DAYS_TIMEDELTA = datetime.timedelta(days=60)

holidays_us = holidays.US()
upcoming_holiday_dates = holidays_us.holidays_before(SIXTY_DAYS_TIMEDELTA)
holiday_dates = holidays_us.holidays_between(NOW_DATETIME, NOW_DATETIME + SIXTY_DAYS_TIMEDELTA)
```

It's not clear if _holidays_ supports just checking certain dates in order for developers to be able to implement this ourselves. It seems _sort of_ there, except that I wonder what happens due to the year handling if the next sixty days includes both December and January. Based on the design of `_add_holiday()`, they're seemingly stored in a year-specific fashion:
https://github.com/vacanza/python-holidays/blob/fde28db1d17e05e0853585936b8b8f85e9e38d34/holidays/holiday_base.py#L737

Suggestions?

**Comment by arkid15r** [comment]
Hi @dsoprea!

>I'd like to be able to determine if there are any holidays in the near future, expressed as a timedelta. More generally, maybe just between two dates.

I don't recall `holidays` having built-in functionality for that. However, it sounds to me like something that could be useful for the end-users. 

PRs welcome. Or you may get lucky if @KJhellico or me will be able to find some time this month. Perhaps, by extending `holidays.utils.country_holidays()` (and similar) with `before_date` and `after_date` args.

Thanks for the idea!

**Comment by dsoprea** [comment]
If you could suggest a straightforward approach then I could do it. If it requires research or rewiring then that'd be best left to the maintainers.

**Comment by arkid15r** [comment]
Yeah, it's up to you. I believe any PR would require some research of the existing code base, project architecture and generally used approaches. I suggest looking at some of the most recent PRs (perhaps, changes related to `HolidayBase` would be the most useful).

Just make sure to let us know if you decide to work on this so no work would be done twice.

Thank you @dsoprea!

**Comment by Rosi2143** [comment]
Hi @arkid15r / @dsoprea ,

I'm looking for something similar - but would like to revert the logic.

How about adding a function to get the "next" holiday. Getting the amount of days is easy with existing datetime functions.

e.g. a function returning the date.

`
>>> us_holidays = holidays.US()
>>> next_holiday = us_holidays.getNextHoliday()
>>> print(next_holiday)
2025-01-20
`

I would be willing to create a MR if you agree with the approach,

Thanks

P.S. Another idea for an extention would be to get a function to get the "number of working days" (similar to get_nth_working_day) from now to some other date - e.g. the next holiday returned via getNextHoliday().

**Comment by arkid15r** [comment]
> Hi @arkid15r / @dsoprea ,
> 
> I'm looking for something similar - but would like to revert the logic.
> 
> How about adding a function to get the "next" holiday. Getting the amount of days is easy with existing datetime functions.
> 
> e.g. a function returning the date.
> 
> `
> 
> > > > us_holidays = holidays.US()
> > > > next_holiday = us_holidays.getNextHoliday()
> > > > print(next_holiday)
> > > > 2025-01-20
> > > > `
> 
> I would be willing to create a MR if you agree with the approach,
> 
> Thanks
> 
> P.S. Another idea for an extention would be to get a function to get the "number of working days" (similar to get_nth_working_day) from now to some other date - e.g. the next holiday returned via getNextHoliday().

Sounds good, I'd only recommend it to return a tuple 

> print(next_holiday)
> (2025-01-20, "Martin Luther King Jr. Day")

Maybe also add a `dt: date` as an optional parameter for the next holiday calculation.

@KJhellico @PPsyrius any suggestions?

**Comment by Rosi2143** [comment]
OK guys,

I will then give it a try.

And yes - returning a tuple is a good idea. Thanks.

**Comment by KJhellico** [comment]
> More generally, maybe just between two dates

Actually, it's possible to use slice syntax:
```python
>>> from holidays.countries.united_states import US
>>> us_holidays = US(years=2025)
>>> part_holidays = us_holidays["2025-04-01":"2025-07-01"]
>>> print(part_holidays)
[datetime.date(2025, 5, 26), datetime.date(2025, 6, 19)]
```
